### PR TITLE
feat: Provide minimal distribution of CRDs

### LINKED ISSUE #3487: InferenceService CRD is too large
/kind bug

KServe's [ISVC CRD](https://github.com/kserve/kserve/blob/master/config/crd/serving.kserve.io_inferenceservices.yaml) is too large and it's impossible to add new fields to the spec given the limitation of K8s CRD size in general.

This currently blocks a couple of community PRs, including https://github.com/kserve/kserve/pull/3479 and https://github.com/kserve/kserve/pull/3418.

Even though the error message is pointing to the annotations, the actual problem might be the size of the entire CRD. 
`CustomResourceDefinition "[inferenceservices.serving.kserve.io](http://inferenceservices.serving.kserve.io/)" is invalid: metadata.annotations: Too long: must have at most 262144 bytes`

### PR DESCRIPTION
**Pull request by terrytangyuan**
Fixes https://github.com/kserve/kserve/issues/3487.

Trying out a similar approach to what we are doing for Argo Workflows https://github.com/argoproj/argo-workflows

### CODE REVIEW

File: hack/main.go [hunk 1]
```diff
@@ -0,0 +1,14 @@
+package main
```
Comment (yuzisun) [inline hack/main.go]: Can we move these files under a different directory?  we mostly store scripts under `hack`

File: config/crd/full/kustomization.yaml [hunk 2]
```diff
@@ -0,0 +1,10 @@
+apiVersion: kustomize.config.k8s.io/v1beta1
```
Comment (yuzisun) [inline config/crd/full/kustomization.yaml:1]: We also need to update the crd helm chart, probably need to a make a new chart for minimal crd

File: hack/main.go (hunk 1 above)
Comment (terrytangyuan) [reply hack/main.go]: Moved to `cmd/crd-gen`

File: config/crd/full/kustomization.yaml (hunk 2 above)
Comment (terrytangyuan) [reply config/crd/full/kustomization.yaml:1]: Added kserve-crd-minimal chart

**Comment by terrytangyuan** [comment]
@yuzisun Addressed comments. PTAL.

**Comment by terrytangyuan** [comment]
@yuzisun @sivanantha321 Just rebased and resolved conflicts after the merge of https://github.com/kserve/kserve/pull/3443.

File: hack/crdgen.sh [hunk 3]
```diff
@@ -0,0 +1,13 @@
+#!/bin/bash
```
Comment (sivanantha321) [inline hack/crdgen.sh]: This script only generates minimal crd right ? If yes we can name the script more specifically

**Comment by sivanantha321** [comment]
I am just curious, How is the validation handled for minimal crd version. Is it through validation webhook ?

File: hack/crdgen.sh (hunk 3 above)
Comment (terrytangyuan) [reply hack/crdgen.sh]: Renamed!

**Comment by terrytangyuan** [comment]
@sivanantha321 Yes and basically leave that responsibility to something else, e.g. front-end, validation webhook, etc.

**Pull request merged**